### PR TITLE
[FEATURE] Add command controller to create test data

### PR DESCRIPTION
**Pull request by kanow**
Simple command to add test data for tea model with a console command. It's just the first try and should be extended by more data, translations, etc.

Resolves: #1120

### CODE REVIEW

**Comment by sbuerk** [comment]
I know this is WIP, but I wanted to provide an early feedback on this.

Basically, I would recommend to avoid using `QueryBuilder` or `Connection`
methods (low-level api) for such tasks, until absolutly needed. Usually,
it's more recommandable to use `DataHandler` for operation, as it takes
care of a lot of stuff. For example:

* Default value handling (TCA)
* TYPO3 system field handling
* soft-delete/hard-delete of records based on TCA configuration
* record history for changes (soft-delete, creation, updating, ...)
* automatically exeucte hooks configured by extensions/core
* workspace support if workspace is installed etc without the need to handle system fields directly
* system field for localization of records (l18n_diff, l18n_source) etc
* access verification (if not executed with admin permissions)

albeit eventually not having all in place right now.

Thus I highly recommend to switch to `DataHandler` for the deletion
and data creation/update part.

File: Classes/Command/CreateTestDataCommand.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+declare(strict_types=1);
+
+namespace TTN\Tea\Command;
+
+/*
+ * Command to create test data for the tea extension.
+ */
+
+use phpDocumentor\Reflection\Types\Integer;
```
Comment (sbuerk) [inline Classes/Command/CreateTestDataCommand.php:11]: I assume that this is wrong - what's the reason to use the phpDocumentor reflection type here ? Maybe some error with auto import ?

Comment (kanow) [reply Classes/Command/CreateTestDataCommand.php]: Yes, now removed.

**Comment by kanow** [comment]
We are aware to the usage of DataHandler or not. Daniel means that Command Controller should be an simple example to work with database actions. Maybe later we can extend that existing command or create a second command controller that shows an advanced example to work with the DataHandler.
For example adding more data with relations to images.

File: Classes/Command/CreateTestDataCommand.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+declare(strict_types=1);
+
+namespace TTN\Tea\Command;
+
+/*
+ * Command to create test data for the tea extension.
+ */
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:9]: This class comment should go directly above the class.

File: Classes/Command/CreateTestDataCommand.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use TYPO3\CMS\Core\Database\Connection;
+use TYPO3\CMS\Core\Database\ConnectionPool;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+final class CreateTestDataCommand extends Command
+{
+    /** @phpstan-ignore-next-line */
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:22]:
```suggestion
    /**
     * @var list<array{title: non-empty-string, description: non-empty-string, sys_language_uid: int<0, max>}>
     */
```

File: Classes/Command/CreateTestDataCommand.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            'description' => 'I love that tea!',
+            'sys_language_uid' => 0,
+        ],
+         [
+            'title' => 'Earl Grey',
+            'description' => 'A nice tea!',
+            'sys_language_uid' => 0,
+        ],
+    ];
+    protected function configure(): void
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:35]:
```suggestion

    protected function configure(): void
```

File: Classes/Command/CreateTestDataCommand.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $this
+            ->setHelp('Create test data for the tea extension in an already existing page (sysfolder).')
+            ->addArgument(
+                'pageId',
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:40]: This probably should be `pageUid`, not `pageId`.

File: Classes/Command/CreateTestDataCommand.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                'delete-data-before',
+                'd',
+                InputOption::VALUE_NONE,
+                'Delete all tea data in the defined pid before creating new data.'
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        /** @var int $pageId */
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:54]: In production code, please let's use `assert` to state data restrictions that we (as human) know to be true.

File: Classes/Command/CreateTestDataCommand.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        $pageId = $input->getArgument('pageId') ?? 0;
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:55]: `pageUid`

File: Classes/Command/CreateTestDataCommand.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        /** @var bool $deleteDataBefore */
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:56]: Ditto about `assert`.

File: Classes/Command/CreateTestDataCommand.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        $deleteDataBefore = $input->getOption('delete-data-before') ?? false;
+        $table = 'tx_tea_domain_model_tea';
+        $connectionForTable = GeneralUtility::makeInstance(ConnectionPool::class)->getConnectionForTable($table);
+
+        if ($deleteDataBefore) {
+            $query = $connectionForTable;
+            $query->delete($table, ['pid' => $pageId], [Connection::PARAM_INT]);
+            $output->writeln(sprintf('Existing data in page %s deleted.', $pageId));
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:64]: Please let's use the `DataHandler` for all delete and insert operations so indices and cache clearing are used. Feel free to steal from here: https://github.com/oliverklee/ext-seminars/blob/main/Classes/Domain/Repository/Event/EventRepository.php#L291

File: Classes/Command/CreateTestDataCommand.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use TYPO3\CMS\Core\Database\Connection;
+use TYPO3\CMS\Core\Database\ConnectionPool;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+final class CreateTestDataCommand extends Command
```
Comment (oliverklee) [inline Classes/Command/CreateTestDataCommand.php:20]: Please let's have functional tests for this command.

**Review by oliverklee — CHANGES_REQUESTED** [review]

**Review by oliverklee — CHANGES_REQUESTED** [review]
Currently, the functional tests are failing. Also, we currently have some style warnings and PHPStan warnings which we'll need to fix first.

**Review by oliverklee — CHANGES_REQUESTED** [review]
Currently, some tests are red. Could you please fix these and also rename the Composer script key as requested in an earlier review? Thanks! :pray:

**Comment by kanow** [comment]
Oh, shit. And yes, I still try to make this PR ready to review/merge ;-).

**Review by oliverklee — APPROVED** [review]

**Pull request merged**